src/main/scala/lms/transformation/distributed_tensor/distributedTensorBase.scala

```diff
@@ -0,0 +1,212 @@
+  }
+
+  class TENSOR(override val x: Backend.Exp, val useOldMetadata: Boolean = false) extends TOP(x) {
+    def withEleType(m: Manifest[_]): this.type = { Adapter.typeMap(x) = m; this }
+    override def withSrcType(pos: SourceContext, m: Manifest[_]): this.type =
+      withSource(pos).withEleType(m)
+
+    def pos: SourceContext = if (useOldMetadata) Adapter.oldSourceMap(x) else Adapter.sourceMap(x)
+
+    def et: Manifest[_] = if (useOldMetadata) Adapter.oldTypeMap(x) else Adapter.typeMap(x)
+
+    def tensor_type: TensorType = {
+      val gc = if (useOldMetadata) Adapter.oldDefsCache else Adapter.g.globalDefsCache
+      gc.get(x.asInstanceOf[Backend.Sym]) match {
+        case Some(Node(_, s, Backend.Const(tt:TensorType)::_, _)) if s.startsWith("tensor") => tt
+        case a => throw new Exception(s"cannot find node $a")
+      }
+    }
+
+    def annotation: Anno = {
+      val gc = if (useOldMetadata) Adapter.oldDefsCache else Adapter.g.globalDefsCache
+      gc.get(x.asInstanceOf[Backend.Sym]) match {
+        case Some(Node(_, s, tt::Backend.Const(a:Anno)::_, _)) if s.startsWith("tensor") => a
+        case a => throw new Exception(s"cannot find node $a")
+      }
+    }
+
+    def shape_size: Seq[Int] = tensor_type.shape.map(_.size)
+
+    def show(implicit __pos: SourceContext): UNIT = {
+      UNIT(Adapter.g.reflectEffect("show_tensor", x)(x)(Adapter.CTRL))
+    }
+
+    // FIXME(feiw) save to where?
+    def save(implicit __pos: SourceContext): UNIT = {
+      UNIT(Adapter.g.reflectEffect("save", x)(x)(Adapter.CTRL))
+    }
+  }
+
+  // FIXME(feiw) the return type fo MODULE is strange
+  def MODULE(f: () => TENSOR)(implicit __pos: SourceContext): () => UNIT = {
+    val m = Adapter.g.reflectWrite("module", Adapter.g.reify(f().x))(Adapter.CTRL)
+    () => UNIT(Adapter.g.reflectWrite("@", m, Backend.Const(()))(Adapter.CTRL))
+  }
+
+  def mergable_dims(node: Node): List[(Dim, Dim)] = node match {
+    case Node(s, op, _, _) if (op == "tensor_input" || op == "tensor_weight") => List()
+    case Node(s, op, _, _) =>
+      assert(!op.startsWith("tensor"), s"node $node is not yet handled in mergable_dims")
+      List()
+  }
+
+  def aircopCollect(node: Node, forwardNodes: mutable.ArrayBuffer[Node],
+      weightNodes: mutable.ArrayBuffer[Node], backwardNodes: mutable.ArrayBuffer[()=>Unit],
+      grad_map: mutable.HashMap[Backend.Sym, TENSOR],
+      momentum_map: mutable.HashMap[Backend.Sym, TENSOR],
+      transform: Backend.Exp => Backend.Exp): Unit = node match {
+
+    case Node(s, "tensor_input", _, _) => forwardNodes += node
+    case Node(s, "tensor_weight", _, _) => weightNodes += node
+    case Node(s, op, _, _) => throw new Exception(s"op $op is not yet handled in aircopCollect")
+  }
+}
+
+
+trait FixedSizeDistributedTensorOpsBase extends Dsl {
+
+  import PrimitiveTypeLess._
+  import FixedSizeDistributedTensorTypeLess._
+
+  def module[T](f: () => Rep[Tensor[T]])(implicit __pos: SourceContext) = {
```

Comment:
why the unit argument is necessary? it looks like `f` can be just passed by name `=> Rep[Tensor[T]]`